### PR TITLE
New: `no-useless-concat` rule (fixes #3506)

### LINKED ISSUE #3506: no-useless-concat rule
Prefer-Template doesn't detect adding 2 template string
If you have the code
```js
const a = `some` + `string`;
```
It won't be detected by the rule

### PR DESCRIPTION
**Pull request by hzoo**
https://github.com/eslint/eslint/issues/3506

WIP

@mysticatea @Cellule

### CODE REVIEW

**Comment by mysticatea** [comment]
Oh very quickly!

Could you add a test `var a = 1 + 1;` and `var a = 1 * "2";` into valid?
And I prefer warning at `+` operator.
If it's at the head of BinaryExpression, `"a" + "b" + "c"` looks duplicated warning.

**Comment by hzoo** [comment]
K i'l add those tests

> And I prefer warning at + operator.
> If it's at the head of BinaryExpression, "a" + "b" + "c" looks duplicated warning.

Sorry, Can you reiterate what you mean here?

**Comment by mysticatea** [comment]
I had guessed it happens below:

```
> eslint test.js

test.js
  1:1  error  Unexpected string concatenation of literals.  no-useless-concat
  1:1  error  Unexpected string concatenation of literals.  no-useless-concat
```

But I was wrong. A part of `"b" + "c"` is a concatenation between BinaryExpression and Literal.

**Comment by hzoo** [comment]
Ok added a case to check if both left and right were numbers

**Comment by hzoo** [comment]
@mysticatea Oh ok I see what you mean - I figured that out as well from writing the other rule.

**Comment by Cellule** [comment]
This looks good to me

**Comment by mysticatea** [comment]
Well, I had a rethink on `foo + "a" + "b"`.

This rule can be used for ES3/5 code. 
So I think this rule should warn `foo + "a" + "b"` (BinaryExpression + Literal/TemplateLiteral).

```js
function getLeft(node) {
    var left = node.left;
    while (left.type === "BinaryExpression" && left.operator === "+") {
        left = left.right;
    }
    return left;
}
```

And to warn at `+` operator:

```js
var opToken = context.getTokenAfter(left);
while (opToken.value !== "+") {
    opToken = context.getTokenAfter(opToken);
}
context.report(node, opToken.loc.start, "...");
```

I'm not sure, this way may work....

**Comment by hzoo** [comment]
Ok added `getLeft` to account for `foo + "a" + "b"`

@mysticatea Can you explain the warning for the 2nd thing you said?

**Comment by mysticatea** [comment]
Thank you.

The 2nd thing is to avoid:

```
"a" + "b" + "c"
```

```
> eslint test.js

test.js
  1:1  error  Unexpected string concatenation of literals.  no-useless-concat
  1:1  error  Unexpected string concatenation of literals.  no-useless-concat
```

If we give an `{line: number, column: number}` object to the 2nd argument of `context.report()`, we can move the warning location. ([semi](https://github.com/eslint/eslint/blob/master/lib/rules/semi.js#L25-L38) does the same thing)
It move the warning location to its `+` operator. (Skipping non `+` to support `("a" + "b") + "c"`)

```
> eslint test.js

test.js
  1:5   error  Unexpected string concatenation of literals.  no-useless-concat
  1:11  error  Unexpected string concatenation of literals.  no-useless-concat
```

Maybe the location of `left` node is good as well.

**Comment by hzoo** [comment]
Ok I see you meant the location warned should change - done?

**Comment by mysticatea** [comment]
Looks good to me.
I'm thankful for your patience.

**Comment by hzoo** [comment]
k anyone else? @nzakas

File: docs/rules/no-useless-concat.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Disallow unncessary concatenation of strings (no-useless-concat)
+
+It's unncessary to concatenate two strings together since they can be combined into a single string.
```
Comment (ilyavolodin) [inline docs/rules/no-useless-concat.md]: I would reword it so that the meaning would imply that concatenating two strings on a single line is not necessary. I think right now the statement is too broad.

File: lib/rules/no-useless-concat.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+module.exports = function(context) {
+    return {
+        BinaryExpression: function(node) {
+            // account for the `foo + "a" + "b"` case
+            var left = getLeft(node);
+            var right = node.right;
+
+            if ((left.type === "Literal" || left.type === "TemplateLiteral") &&
+                (right.type === "Literal" || right.type === "TemplateLiteral") &&
+                node.operator === "+"
```
Comment (ilyavolodin) [inline lib/rules/no-useless-concat.js]: Can you move this as a first check to short-circuit anything other then concatenation?

**Comment by hzoo** [comment]
@ilyavolodin k hopefully that's better?

**Comment by ilyavolodin** [comment]
LGTM, thanks!
